### PR TITLE
Katana to eight

### PR DESCRIPTION
**Pull request by mariangemarcano**
Katana 2.0 Release:
- Feature: Run hg / Find Previous Successful Build / Check if Artifact Exists / Trigger steps on different set of slaves which are available for this task.
- Feature: Resume build on next available slave (instead of the originally-assigned Slave):Builds has new resume result (json api update). Merging logic handles builds on a resume state.  Cancel build and cancel build chain takes into account builds waiting to be resumed.
- Fix: Adding identify mode to mercurial command, it runs the initial mercurial install check, then performs hg --traceback identify [repourl] --debug --rev [branch or revision], and updates build sourcestamps if the revision was correctly identified. 
- Feature: Add support to run any command at the beginning and end of every build, we will use it for clean up scripts (purify).
- Fix: Trigger step will retry the builds after slave or buildmaster disconnection.
- Optimization: Trigger step check if buildrequests were already inserted on the db when retrying the same build after master or slave disconnection.
- Fix: Handle client disconnected or crashed exceptions, so the master run the callbacks and retry the build again. This error that causes the slave to stay in PINGING state when pinging fails. When we hit this problem the slave is not able to pick up any new jobs once it becomes available again.
- Feature: Expose via json when slave is paused or grateful shut down.
- Feature: Expose via json api the number of builds that katana run during the last day.
- Fix: sqlalchemy OperationalError Deadlock when trying to update buildrequests.
- Fix: Trigger step failure not valid scheduler: error when retrying builds after buildmaster restart. 
  Features disabled: Calculate incoming changes feature, BuildQueue page.
- Fix: Show a message when the requested slave is not currently attached to the master.
- Fix: self.updateHealth() was not running the async version of generate finished builds, this blocked the buildmaster from running any tasks.
  Feature: New katana 2.0 ui, katana front page bookmarks the history of last 20 project / builders page visited.
- Feature: Merged several fixes and improvements from upstream. Some features include:
- Feature: Ability to pause slave.
- Feature: Merged several optimizations and refactoring around the way buildbot process buildrequest. Now, takes locks into account, and will not schedule a build only to have it wait on a lock, it also introduces a canStartBuild builder configuration option which can be used to prevent a build request being assigned to a slave.

### CODE REVIEW

File: master/buildbot/process/build.py [hunk 1]
```diff
@@ -295,38 +300,85 @@ def acquireLocks(self, res=None):
     def _startBuild_2(self, res):
         self.startNextStep()
 
-    def setupBuild(self, expectations):
-        # create the actual BuildSteps. If there are any name collisions, we
-        # add a count to the loser until it is unique.
-        self.steps = []
-        self.stepStatuses = {}
+    def setupStep(self, factory, stepnames):
+        step = factory.buildStep()
```
Comment (opdude) [inline master/buildbot/process/build.py]: Lets rename stepnames to something that makes more sense i.e step_counts

File: master/buildbot/process/build.py [hunk 2]
```diff
@@ -560,7 +626,9 @@ def buildFinished(self, text, results):
             # XXX: also test a 'timing consistent' flag?
             log.msg(" setting expectations for next time")
             self.builder.setExpectations(self.progress)
-        reactor.callLater(0, self.releaseLocks)
+        # TODO: check why callbacks are not triggered when stopping master and retrying builds
+        # this maybe cause by the upstream merged
+        eventually(self.releaseLocks)
```
Comment (opdude) [inline master/buildbot/process/build.py]: What implications will this have for us? Might we miss data going to bfstats?

Comment (mariangemarcano) [inline master/buildbot/process/build.py]: that was introduced by upstream, to handle the call queue and behind the scenes it calls the reactor.callLater from twisted, we adjusted the stats and tested with this changed and they are working fine.

File: master/buildbot/process/buildrequestdistributor.py [hunk 3]
```diff
@@ -0,0 +1,1004 @@
...
+        defer.returnValue(self.unclaimedBrdicts)
+
+    @defer.inlineCallbacks
+    def claimBuildRequests(self, breqs):
+        brids = [br.id for br in breqs]
+        yield self.master.db.buildrequests.claimBuildRequests(brids)
+    
+    @defer.inlineCallbacks
+    def _getBuildRequestForBrdict(self, brdict):
+        # Turn a brdict into a BuildRequest into a brdict. This is useful
```
Comment (opdude) [inline master/buildbot/process/buildrequestdistributor.py]: Turn a brdict into a BuildRequest and back again?

Comment (mariangemarcano) [inline master/buildbot/process/buildrequestdistributor.py]: @opdude the comment seems wrong it creates a BuildRequest from a brdict and this object is cached

File: master/buildbot/process/buildrequestdistributor.py [hunk 4]
```diff
@@ -0,0 +1,1004 @@
...
+    calling their C{maybeStartBuild} method.
+
+    This takes account of the C{prioritizeBuilders} configuration, and is
+    highly re-entrant; that is, if a new build request arrives while builders
+    are still working on the previous build request, then this class will
+    correctly re-prioritize invocations of builders' C{maybeStartBuild}
+    methods.
+    """
+
+    # todo read from configuration if want to use buildbot's or katana build chooser
```
Comment (opdude) [inline master/buildbot/process/buildrequestdistributor.py]: change todo to TODO: so it shows up as a todo in pycharm :)

**Comment by opdude** [comment]
Looks fine, can you make cards for these minor issues?

**Comment by mariangemarcano** [comment]
@opdude sure

**Comment by elmirjagudin** [comment]
"Feature: Add support to run any command at the beginning and end of every build, we will use it for clean up scripts (purify)."

Can't we use a normal buildstep for this?

**Comment by mariangemarcano** [comment]
@elmirjagudin yes we can but we want it to make it global so we don't need to add it every where on each factory right now we have more that 500 build factories

File: master/buildbot/process/buildrequestdistributor.py [hunk 5]
```diff
@@ -0,0 +1,1004 @@
...
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+
```
Comment (elmirjagudin) [inline master/buildbot/process/buildrequestdistributor.py]: There should be a section in the developer documentation which describes the high level algorithm used for scheduling builds. How the slaves are picked, how the builds are prioritized, etc.

And we should add a reference to that section in this file.

**Pull request merged**